### PR TITLE
CODENVY-619 Restrict usernames to letters and digits only

### PR DESCRIPTION
**Pull request by mkuznyetsov**
Users with names, that are not URL-friendly must not be created

### CODE REVIEW

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java [hunk 1]
```diff
@@ -429,6 +434,10 @@ public UserDescriptor getByName(@ApiParam(value = "User email")
         return ImmutableMap.of(USER_SELF_CREATION_ALLOWED, Boolean.toString(userSelfCreationAllowed));
     }
 
+    private boolean isValidUserName(String name) {
+        return Pattern.matches("^[a-zA-Z0-9]*$", name);
```
Comment (sleshchenko) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: mb it is better to compile Patter once and store it to static field

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java [hunk 2]
```diff
@@ -429,6 +436,10 @@ public UserDescriptor getByName(@ApiParam(value = "User email")
         return ImmutableMap.of(USER_SELF_CREATION_ALLOWED, Boolean.toString(userSelfCreationAllowed));
     }
 
+    private boolean isValidUserName(String name) {
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: inline

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java [hunk 3]
```diff
@@ -50,10 +54,12 @@
     @Inject
     public UserManager(UserDao userDao,
                        UserProfileDao profileDao,
-                       PreferenceDao preferenceDao) {
+                       PreferenceDao preferenceDao,
+                       @Named("account.name.reserved_names")String[] reservedNames) {
```
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java]: probably missing space after names")

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java [hunk 4]
```diff
@@ -38,7 +42,7 @@
  * @author Max Shaposhnik (mshaposhnik@codenvy.com)
  */
 public class UserManager {
-
+    List<String> reservedNames = new ArrayList<>();
```
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java]: missing java modifier (probably private)
also we don't need to initialize variable to a list as in constructor we assign the value

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java [hunk 5]
```diff
@@ -161,6 +165,9 @@ public Response create(@ApiParam(value = "New user")
         }
 
         final User user = context.isUserInRole("system/admin") ? fromEntity(userDescriptor) : fromToken(token);
+        if (isValidUserName(user.getName())) {
```
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: it looks odd to have isValid call that throw exception ? like there is a problem there

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java [hunk 6]
```diff
@@ -84,6 +91,15 @@ public void create(User user, boolean isTemporary) throws ConflictException, Ser
         }
     }
 
+    private void validateUserName(String name) throws ConflictException {
+
+        for (String reservedName : reservedNames) {
+            if (name.equals(reservedName)) {
```
Comment (benoitf) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java]: could be better to use reserverName as left operand

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java (hunk 3 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java]: removed

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java (hunk 4 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java]: removed

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java (hunk 5 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: removed

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java (hunk 6 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserManager.java]: removed

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java (hunk 2 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: removed

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java (hunk 1 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: done

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
+package org.eclipse.che.api.user.server;
```
Comment (sleshchenko) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java]: missed license

**Comment by skabashnyuk** [comment]
ok.  @evoevodin ?

File: wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    public void testNormalizeUserName(String input, String expected) {
+
+        Assert.assertEquals(validator.normalizeUserName(input), expected);
+    }
+
+
+    @Test(dataProvider = "validNames")
+    public void testValidUserName(String input, String expected) {
+
+        Assert.assertEquals(validator.normalizeUserName(input), expected);
```
Comment (sleshchenko) [inline wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java]: In fact this test does the same as testNormalizeUserName, no?

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.regex.Pattern;
+
+/**
+ * Utils for username validation and normalization.
+ *
+ * @author Mihail Kuznyetsov
+ */
+public class UserNameValidator {
+    private static final Pattern ILLEGAL_USERNAME_CHARACTERS = Pattern.compile("[^a-zA-Z0-9]");
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java:21]: We can allow the "_", "-" characters. They are url allowed.

File: wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserServiceTest.java [hunk 10]
```diff
@@ -34,6 +34,7 @@
 import org.mockito.Mock;
 import org.mockito.testng.MockitoTestNGListener;
 import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
```
Comment (sleshchenko) [inline wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserServiceTest.java]: unused import?

File: wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    @BeforeMethod
+    public void setUp() {
+        validator = new UserNameValidator();
+    }
+
+    @Test(dataProvider = "normalizeNames")
+    public void testNormalizeUserName(String input, String expected) {
+
+        Assert.assertEquals(validator.normalizeUserName(input), expected);
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java]: why static methods called as instance methods ?

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java (hunk 9 above)
Comment (skabashnyuk) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java:21]: In future we can do like github. Allow "-" in the middle.

File: wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserServiceTest.java (hunk 10 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserServiceTest.java]: fixed

File: wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java (hunk 11 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java]: fixed

File: wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java (hunk 8 above)
Comment (mkuznyetsov) [inline wsmaster/che-core-api-user/src/test/java/org/eclipse/che/api/user/server/UserNameValidatorTest.java]: fixed

**Comment by skabashnyuk** [comment]
ок

**Comment by sleshchenko** [comment]
OK for me

**Pull request merged**